### PR TITLE
V2.0.0 alpha.6

### LINKED ISSUE #61: upgrade to v2.0.0-alpha.5
this is rather involved. We need to support generic hashers, as defined in metadata:
```
       {
        "name": "Account",
        "modifier": "Default",
        "ty": {
         "Map": {
          "hasher": "Blake2_128Concat",
          "key": "T::AccountId",
          "value": "AccountInfo<T::Index, T::AccountData>",
          "unused": false
         }
        },

```

And for double_map, the two hashers can be different

```
       {
        "name": "MeetupRegistry",
        "modifier": "Default",
        "ty": {
         "DoubleMap": {
          "hasher": "Blake2_256",
          "key1": "CurrencyCeremony",
          "key2": "MeetupIndexType",
          "value": "Vec<T::AccountId>",
          "key2_hasher": "Blake2_256"
         }
        },

```

### PR DESCRIPTION
**Pull request by brenzi**
* upgrading
* deprecating hard-to-maintain samples

fixes #61 

* now building against a [fork of substrate-node-template](https://github.com/scs/substrate-node-template/tree/test-api-client)
* storage map and double map working. extended example 
* minor refactoring
* includes new api for get_block aso.
* works for v2.0.0-alpha.5 based [encointer-client](https://github.com/encointer/encointer-node/tree/master/client)

### technical debt
* node template has no kitties and no contracts. do we really want to maintain substrate-test-nodes? We could just maintain a branch on a fork of node-template, like I started to do (see above)
* unsure, if using AccountId32 is a good idea. I think we should use the Runtime itself a a type argument to Api, like subxt does. then we can derive `Runtime::AccountId`

### CODE REVIEW

**Comment by brenzi** [comment]
@clangenb please check if this PR includes all your previous work

File: src/extrinsic/xt_primitives.rs [hunk 1]
```diff
@@ -21,12 +21,16 @@ use sp_std::prelude::*;
...
+use sp_core::H256;
 use sp_core::blake2_256;
 use sp_runtime::{generic::Era, MultiSignature};
-pub type GenericAddress = Address<AccountId, AccountIndex>;
+
+pub use sp_runtime::AccountId32 as AccountId;
+
+pub type AccountIndex = u64;
+
+pub type GenericAddress = AccountId; //Address<AccountId, AccountIndex>;
```
Comment (clangenb) [inline src/extrinsic/xt_primitives.rs:33]: why do that redundant rename? I think it would be better to use AccountId

File: src/lib.rs [hunk 2]
```diff
@@ -101,7 +117,7 @@ where
 
         let meta = Self::_get_metadata(url.clone());
         let metadata = Metadata::try_from(meta).unwrap();
-        info!("Metadata: {:?}", metadata);
+        debug!("Metadata: {:?}", metadata);
```
Comment (clangenb) [inline src/lib.rs:120]: really happy about that!

File: src/lib.rs [hunk 3]
```diff
@@ -197,62 +181,143 @@ where
...
-                arr.clone_from_slice(key.to_owned().public().as_ref());
-                Ok(Self::_get_nonce(self.url.clone(), arr))
+                arr.clone_from_slice(pair.to_owned().public().as_ref());
+                let accountid: AccountId = Decode::decode(&mut &arr.encode()[..]).unwrap();
+                if let Some(info) = self.get_account_info(&accountid) {
+                    Ok(info.nonce)
+                } else { Ok(0) }
             }
-            None => Err("Can't get nonce when no signer is set"),
+            None => Err("Can't get )nonce when no signer is set"),
```
Comment (clangenb) [inline src/lib.rs:192]: typo

File: src/node_metadata.rs [hunk 4]
```diff
@@ -265,12 +320,14 @@ impl<K: Encode, V: Decode + Clone> StorageMap<K, V> {
         bytes.extend(&sp_core::twox_128(&self.storage_prefix)[..]);
         let encoded_key = key.encode();
         let hash = match self.hasher {
+            StorageHasher::Identity => encoded_key.to_vec(),
             StorageHasher::Blake2_128 => sp_core::blake2_128(&encoded_key).to_vec(),
             StorageHasher::Blake2_128Concat => {
                 // copied from substrate Blake2_128Concat::hash since StorageHasher is not public
-                sp_core::blake2_128(&encoded_key)
+                let x: &[u8] = &encoded_key[..];
```
Comment (clangenb) [inline src/node_metadata.rs:327]: you could as well use `encoded_key.as_slice()` , which I personally prefer

File: src/node_metadata.rs [hunk 5]
```diff
@@ -288,6 +345,49 @@ impl<K: Encode, V: Decode + Clone> StorageMap<K, V> {
...
+    pub fn key(&self, key1: K, key2: Q) -> StorageKey {
+        let mut bytes = sp_core::twox_128(&self.module_prefix).to_vec();
+        bytes.extend(&sp_core::twox_128(&self.storage_prefix)[..]);
+        for (encoded_key, hasher) in [(key1.encode(), self.hasher.clone()), (key2.encode(), self.key2_hasher.clone())].iter() {
+            let hash = match hasher {
+                StorageHasher::Identity => encoded_key.to_vec(),
+                StorageHasher::Blake2_128 => sp_core::blake2_128(&encoded_key).to_vec(),
+                StorageHasher::Blake2_128Concat => {
+                    // copied from substrate Blake2_128Concat::hash since StorageHasher is not public
+                    let x: &[u8] = &encoded_key[..];
```
Comment (clangenb) [inline src/node_metadata.rs:369]: why make x?

**Comment by clangenb** [comment]
and yes, it does include my work

File: src/extrinsic/xt_primitives.rs (hunk 1 above)
Comment (brenzi) [reply src/extrinsic/xt_primitives.rs:33]: yes, but the problem is that we will want to get back to generic in order to support polkadot/kusama  chains as well which have a different definition.

File: src/node_metadata.rs (hunk 4 above)
Comment (brenzi) [reply src/node_metadata.rs:327]: ok. was copy-paste from upstream, but I agree

File: src/node_metadata.rs (hunk 5 above)
Comment (brenzi) [reply src/node_metadata.rs:369]: ???
anyway. will replace this redundant code by a fn call

**Pull request merged**